core/src/main/scala/scalaz/Foldable.scala

```diff
@@ -364,15 +365,13 @@ trait Foldable[F[_]]  { self =>
    * Selects groups of elements that satisfy p and discards others.
    */
   def selectSplit[A](fa: F[A])(p: A => Boolean): List[NonEmptyList[A]] =
-    foldRight(fa, (List[NonEmptyList[A]](), false))((a, xb) => xb match {
-      case (x, b) => {
-        val pa = p(a)
-        (if (pa)
-          if (b)
-            (a <:: x.head) :: x.tail else
-            NonEmptyList(a) :: x
-        else x, pa)
-      }
+    foldRight(fa, (List.empty[NonEmptyList[A]], true))((a, lb) => (lb, p(a)) match {
+      case ((l, _), false) =>
+        (l, true)
+      case ((x :: xs, false), true) =>
+        ((a <:: x) :: xs, false)
+      case ((l, _), true) =>
+        (NonEmptyList(a) :: l, false)
```

Comment:
This last case (correctly) assumes that `((Nil, false), true)` never occurs. The previous implementation (correctly) assumed that the unsafe methods were in fact used safely. Should the assumption not hold for any reason, the previous implementation would at least fail fast with an exception. This, however, silently produces wrong result.

I suggest to write out the two cases, `((l, true), true)` and `((Nil, false), true)`, calling `sys.error` in the latter.

Reply:
Actually, I have thought that current implementation is more resilient, since if the assumption does not hold for any reason, then it creates new subNel with current element (there is no other way, because every element which satisfies the predicate, has to be in the result), and continues appending to it based on the predicate.
If you think, that's not the case, then I will change it to `sys.error`

Reply:
I see.

It is somewhat confusing then that it does not matter whether you start with `true` or `false`.

An auxiliary type like

```scala
sealed trait Acc {
  def add(a: A): Acc = ???
  def closeRun: Acc = ???
}
case class InsideRun(currentRun: NonEmptyList[A], tail: List[NonEmptyList[A]]) extends Acc
case class BetweenRuns(tail: List[NonEmptyList[A]]) extends Acc
```

would eliminate the ambiguity. If you find it an overkill, just leave it as is.

Reply:
actually we could write something like:
```  def selectSplit[A](fa: F[A])(p: A => Boolean): List[NonEmptyList[A]] =
    foldRight(fa, (List.empty[NonEmptyList[A]], List.empty[A]))((a, lb) =>
      if (p(a)) lb match {
        case (acc, l) =>
          (acc, a :: l)
      } else (scalaz.std.list.toNel(lb._2).toList ::: lb._1, List.empty)
    )._1.reverse```
This way we accumulate values in `List`, until the predicate fails. When it fails, we prepend this List to the result, and repeat procedure with empty list.
This way semantics of the accumulator stays clear and we do not need to use confusing boolean for tracking.